aws/s3.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	slog.Debug("Downloaded file", "size", numBytes)
+
+	return nil
+}
+
+// ParseAmazonS3URL parses an HTTP/HTTPS URL for an S3 resource and returns an
+// S3URL object.
+//
+// S3 URLs come in two flavors: virtual hosted-style URLs and path-style URLs.
+// Virtual hosted-style URLs have the bucket name as the first component of the
+// hostname, e.g.
+//
+//	https://mybucket.s3.us-east-1.amazonaws.com/a/b/c
+//
+// Path-style URLs have the bucket name as the first component of the path, e.g.
+//
+//	https://s3.us-east-1.amazonaws.com/mybucket/a/b/c
+func ParseAmazonS3URL(s3URL *url.URL) (S3URL, error) {
+	output, err := parseBucketAndRegionFromHost(s3URL.Host)
+	if err != nil {
+		return S3URL{}, errors.NewError("parsing host failed", err, false)
+	}
+
+	output.IsPathStyle = output.Bucket == ""
+
+	path := s3URL.Path
+
+	if output.IsPathStyle {
+		// no bucket name in the authority, parse it from the path
+		output.IsPathStyle = true
+
+		// grab the encoded path so we don't run afoul of '/'s in the bucket name
+		if path == "/" || path == "" {
+		} else {
+			path = path[1:]
+			index := strings.Index(path, "/")
+			if index == -1 {
+				// https://s3.amazonaws.com/bucket
+				output.Bucket = path
+				output.Key = ""
+			} else if index == (len(path) - 1) {
+				// https://s3.amazonaws.com/bucket/
+				output.Bucket = strings.TrimRight(path, "/")
+				output.Key = ""
+			} else {
+				// https://s3.amazonaws.com/bucket/key
+				output.Bucket = path[:index]
+				output.Key = path[index+1:]
+			}
+		}
+	} else {
+		// bucket name in the host, path is the object key
+		if path == "/" || path == "" {
+			output.Key = ""
+		} else {
+			output.Key = path[1:]
+		}
+	}
+
+	if strings.EqualFold(output.Region, "external-1") {
+		output.Region = "us-east-1"
+	} else if output.Region == "" {
+		// s3 bucket URL in us-east-1 doesn't include region
+		output.Region = "us-east-1"
+	}
+
+	return output, nil
+}
+
+func parseBucketAndRegionFromHost(host string) (S3URL, error) {
+	result := vpceUrlRegex.FindStringSubmatch(host)
+	if result != nil && len(result) > vpceUrlPatternBucketIdx && len(result) > vpceUrlPatternRegionIdx {
+		return S3URL{
+			EndPoint: result[vpceUrlPatternHostIdx],
+			Bucket:   result[vpceUrlPatternBucketIdx],
+			Region:   result[vpceUrlPatternRegionIdx],
+		}, nil
+	} else {
+		result = nonVpceUrlRegex.FindStringSubmatch(host)
+		if result != nil && len(result) > vpceUrlPatternBucketIdx && len(result) > vpceUrlPatternRegionIdx {
+			return S3URL{
+				Bucket: result[nonVpceUrlPatternBucketIdx],
+				Region: result[nonVpceUrlPatternRegionIdx],
+			}, nil
+		} else {
+			return S3URL{}, errors.NewError("failed to match URL", nil, false)
+		}
+	}
+}
+
+// DownloadFileFromS3 takes an S3 URL and a filePath, downloads the file from s3 and stores it in the filePath.
+func DownloadFileFromS3(ctx context.Context, downloadURL, filePath string) error {
+	parsedURL, err := url.Parse(downloadURL)
+	if err != nil {
+		return err
```

Comment:
This is already added by the lower layer method.

Reply:
It's good practice to add it on all layers \
Keeping context of lower layers is an overhead

Reply:
We are using errors.New() from vcore, which retains the msg. It would be redundant to add the same error message. The end user just consumes the error returned by DownloadFileFromS3() which will precisely contain where it failed along with the message.

Reply:
Cool. Can you confirm this once? I want to confirm if line #171 shows up along with the download URL being parsed.

Reply:
url.Parse() adds the URL in the error msg. Anyhow, I've added context in the error now.